crates/pico_macros/src/memo.rs

```diff
@@ -0,0 +1,136 @@
+use std::hash::DefaultHasher;
+use std::hash::Hash;
+use std::hash::Hasher;
+
+use darling::{ast::NestedMeta, FromMeta};
+use proc_macro::TokenStream;
+use quote::quote;
+use quote::ToTokens;
+use syn::Signature;
+use syn::{
+    parse_macro_input, parse_quote, Error, FnArg, GenericParam, ItemFn, PatType, ReturnType,
+};
+
+#[derive(Debug, FromMeta)]
+struct MemoArgs {
+    #[darling(default)]
+    reference: bool,
+}
+
+pub(crate) fn memo(args: TokenStream, item: TokenStream) -> TokenStream {
+    let ItemFn {
+        sig,
+        vis,
+        block,
+        attrs,
+    } = parse_macro_input!(item as ItemFn);
+
+    let fn_hash = hash(&sig);
+
+    if sig.inputs.is_empty() {
+        return Error::new_spanned(
+            &sig,
+            "Memoized function must have at least one argument (&mut Database)",
+        )
+        .to_compile_error()
+        .into();
+    }
+
+    let db_arg = match &sig.inputs[0] {
+        FnArg::Typed(PatType { pat, .. }) => pat,
+        _ => unreachable!(),
+    };
+
+    let other_args = sig.inputs.iter().skip(1).map(|arg| match arg {
+        FnArg::Typed(PatType { pat, .. }) => pat,
+        _ => unreachable!(),
+    });
+
+    let argument_types = sig.inputs.iter().skip(1).map(|arg| match arg {
+        FnArg::Typed(PatType { ty, .. }) => ty,
+        _ => unreachable!(),
+    });
+
+    let attr_args = match NestedMeta::parse_meta_list(args.into()) {
+        Ok(v) => v,
+        Err(e) => return Error::new_spanned(&sig, e).to_compile_error().into(),
+    };
+
+    let args_ = match MemoArgs::from_list(&attr_args) {
+        Ok(parsed) => parsed,
+        Err(e) => return e.with_span(&sig).write_errors().into(),
+    };
+
+    let (return_type, default) = match &sig.output {
+        ReturnType::Type(_, ty) => (ty.clone(), false),
+        ReturnType::Default => (parse_quote!(()), true),
+    };
+
+    let mut return_expr = quote! {
+        #db_arg.storage()
+            .derived_nodes()
+            .get(&node_id)
+            .expect("value should exist. This is indicative of a bug in Pico.")
+            .value
+            .as_any()
+            .downcast_ref::<#return_type>()
+            .expect("unexpected return type. This is indicative of a bug in Pico.")
+    };
+
+    let mut new_sig = sig.clone();
+
+    if args_.reference && !default {
```

Comment:
I would expect that `#[memo(reference)] fn foo(db) {}` would return `&()` lol

Reply:
haha, I thought it doesn't make sense, but maybe remove this check indeed and simplify the macro